src/lib/bundles/bundles.ts

```diff
@@ -0,0 +1,308 @@
+    bytes = await readBytes(reader, bytes, 16);
+    const tagsLength = byteArrayToLong(bytes.subarray(0, 8));
+    const tagsBytesLength = byteArrayToLong(bytes.subarray(8, 16));
+    bytes = bytes.subarray(16);
+
+    bytes = await readBytes(reader, bytes, tagsBytesLength);
+    const tagsBytes = bytes.subarray(0, tagsBytesLength);
+    const tags =
+      tagsLength !== 0 && tagsBytesLength !== 0
+        ? deserializeTags(Buffer.from(tagsBytes))
+        : [];
+    if (tags.length !== tagsLength) throw new Error("Tags lengths don't match");
+    bytes = bytes.subarray(tagsBytesLength);
+
+    const transform = new Transform({
+      transform(chunk, _, callback) {
+        this.push(chunk);
+        callback();
+      },
+    });
+
+    // Verify signature
+    const signatureData = deepHash([
+      Buffer.from('dataitem'),
+      Buffer.from('1'),
+      Buffer.from(signatureType.toString()),
+      owner,
+      target,
+      anchor,
+      tagsBytes,
+      transform,
+    ]);
+
+    // Get offset of data start and length of data
+    const dataOffset =
+      2 +
+      sigLength +
+      pubLength +
+      (targetPresent ? 33 : 1) +
+      (anchorPresent ? 33 : 1) +
+      16 +
+      tagsBytesLength;
+    const dataSize = length - dataOffset;
+
+    // Stream the data through the transform
+    if (bytes.byteLength > dataSize) {
+      transform.write(bytes.subarray(0, dataSize));
+      bytes = bytes.subarray(dataSize);
+    } else {
+      let skipped = bytes.byteLength;
+      transform.write(bytes);
+      while (dataSize > skipped) {
+        bytes = (await reader.next()).value;
+        if (bytes === undefined) {
+          throw new Error(
+            `Not enough data bytes  expected: ${dataSize} received: ${skipped}`,
+          );
+        }
+
+        skipped += bytes.byteLength;
+
+        if (skipped > dataSize)
+          transform.write(
+            bytes.subarray(0, bytes.byteLength - (skipped - dataSize)),
+          );
+        else transform.write(bytes);
+      }
+      bytes = bytes.subarray(bytes.byteLength - (skipped - dataSize));
+    }
+
+    transform.end();
+
+    if (id !== createHash('sha256').update(signature).digest('base64url'))
+      throw new Error("ID doesn't match signature");
+
+    const Signer = indexToType[signatureType];
+
+    if (!(await Signer.verify(owner, (await signatureData) as any, signature)))
+      throw new Error('Invalid signature');
+
+    items.push({
+      anchor: Buffer.from(anchor).toString('base64url'),
+      dataOffset: offsetSum + dataOffset,
+      dataSize,
+      id,
+      offset: dataItemOffset,
+      owner: Buffer.from(owner).toString('base64url'),
+      ownerOffset,
+      ownerSize: pubLength,
+      sigName,
+      signature: Buffer.from(signature).toString('base64url'),
+      signatureOffset,
+      signatureSize: sigLength,
+      signatureType: signatureType,
+      size: length,
+      tags,
+      target: Buffer.from(target).toString('base64url'),
```

Comment:
nit: we have unit tested implementations for these b64 conversions in `lib/encoding.ts`

Reply:
I'll take a look. Thanks